### PR TITLE
Create tool for generating direct connection codes

### PR DESCRIPTION
**Pull request by vabold**

### CODE REVIEW

File: tools/directconnectioncode/directconnectioncode.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        j = int(j)
+        assert j >= 0 and j < 256, f"IP address octet {i + 1} exceeds accepted range: 0 - 255"
+
+    ip = socket.inet_aton(ip)
+
+    return ip
+
+
+def generate(ip: str, port: int, passcode: int):
+    assert port >= 0 and port < 65536, "Port exceeds accepted range: 0 - 65535"
```
Comment (MikeIsAStar) [inline tools/directconnectioncode/directconnectioncode.py:23]: The lower bound can be increased to 1024.

Comment (vabold) [reply tools/directconnectioncode/directconnectioncode.py]: Resolved in https://github.com/stblr/mkw-sp/pull/449/commits/c29f50559312b0b7f130f3b8e935913ba5ee349e.

File: tools/directconnectioncode/directconnectioncode.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+from argparse import ArgumentParser
+import re
+import socket
+import struct
+
+
+def parse_ip(ip: str) -> str:
+    assert not re.search(
+        "[^0123456789.]", ip), "Unsupported character found in IP address"
```
Comment (riidefi) [inline tools/directconnectioncode/directconnectioncode.py:9]: You need to escape the period character, else it acts like a wildcard

File: tools/directconnectioncode/directconnectioncode.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    octets = ip.split('.')
+    assert len(octets) == 4, "IP address has unexpected octets: accepts x.x.x.x"
+
+    for i, j in enumerate(octets):
+        j = int(j)
+        assert j >= 0 and j < 256, f"IP address octet {i + 1} exceeds accepted range: 0 - 255"
```
Comment (riidefi) [inline tools/directconnectioncode/directconnectioncode.py:16]: Probably meant "expected"?

File: tools/directconnectioncode/directconnectioncode.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    assert passcode >= 0 and passcode < 2048, "Passcode exceeds accepted range: 0 - 2047"
+
+    ip = parse_ip(ip)
+    port = struct.pack(">H", port)
+    passcode = struct.pack(">H", passcode)
+
+    hex_code = passcode + port + ip
+    direct_code = str(struct.unpack(">Q", hex_code)[0])
+
+    direct_code = "0" * (18 - len(direct_code)) + direct_code
```
Comment (riidefi) [inline tools/directconnectioncode/directconnectioncode.py:33]: Might be easier to just use print formatting for a fixed-length hex number and combine numbers with say `(x << 16) + y`

File: tools/directconnectioncode/directconnectioncode.py (hunk 2 above)
Comment (vabold) [reply tools/directconnectioncode/directconnectioncode.py:9]: ❌

File: tools/directconnectioncode/directconnectioncode.py (hunk 3 above)
Comment (vabold) [reply tools/directconnectioncode/directconnectioncode.py]: Resolved in https://github.com/stblr/mkw-sp/pull/449/commits/42561ce81e698eba7e8e7e4dfd7df60784d67e4c.

**Review by stblr — APPROVED** [review]

**Pull request merged**